sw/lib/sw/device/silicon_creator/rom_patch.c

```diff
@@ -0,0 +1,234 @@
+
+enum {
+  kOtpBase = TOP_DARJEELING_OTP_CTRL_CORE_BASE_ADDR,
+  kSwConfig = OTP_CTRL_SW_CFG_WINDOW_REG_OFFSET,
+};
+
+enum {
+  kRomPatchBaseAddr = OTP_CTRL_PARAM_ROM_PATCH_OFFSET,
+  kRomPatchMaxAddr = kRomPatchBaseAddr + OTP_CTRL_PARAM_ROM_PATCH_SIZE,
+  kRomPatchRegionHeaderSizeBytes = sizeof((struct rom_patch){0}.header),
+  kRomPatchRegionMatchTableSizeBytes = sizeof((struct rom_patch){0}.table),
+  kRomPatchRegionSignatureSizeBytes = kSigVerifyRsaNumBytes,
+  /* Preamble is the patch header and the complete match table. */
+  kRomPatchRegionPreambleSizeBytes =
+      kRomPatchRegionHeaderSizeBytes + kRomPatchRegionMatchTableSizeBytes,
+};
+
+#define PATCH_REGION_CODE_SIZE_BYTES(p)                      \
+    (bitfield_field32_read(p->header, kRomPatchSize) << 2) -            \
+     kRomPatchRegionHeaderSizeBytes -                                   \
+     kRomPatchRegionMatchTableSizeBytes -                               \
+     kRomPatchRegionSignatureSizeBytes)
+
+static inline bool rom_patch_lock_valid(uint32_t patch_header) {
+  return bitfield_field32_read(patch_header, kRomPatchLockValid) ==
+         kMultiBitBool4True;
+}
+
+static inline uint8_t rom_patch_revision(uint32_t patch_header) {
+  return (uint8_t)bitfield_field32_read(patch_header, kRomPatchRevision);
+}
+
+static inline uint16_t rom_patch_size_words(uint32_t patch_header) {
+  return (uint16_t)bitfield_field32_read(patch_header, kRomPatchSize);
+}
+
+static inline size_t rom_patch_code_size_words(const rom_patch_t *patch) {
+  HARDENED_CHECK_NE(patch, NULL);
+
+  return ((bitfield_field32_read(patch->header, kRomPatchSize) << 2) -
+          kRomPatchRegionHeaderSizeBytes - kRomPatchRegionMatchTableSizeBytes -
+          kRomPatchRegionSignatureSizeBytes) >>
+         2;
+}
+
+static inline size_t rom_patch_region_enabled(
+    const rom_patch_match_regs_t *match) {
+  HARDENED_CHECK_NE(match, NULL);
+  return match->m_base != 0 && match->r_base != 0;
+}
+
+static inline bool rom_patch_region_locked(
+    const rom_patch_match_regs_t *match) {
+  HARDENED_CHECK_NE(match, NULL);
+  return bitfield_field32_read(match->m_base, kRomPatchMatchLocked) == 1;
+}
+
+static inline size_t rom_patch_region_size(
+    const rom_patch_match_regs_t *match) {
+  HARDENED_CHECK_NE(match, NULL);
+  return bitfield_field32_read(match->m_base, kRomPatchMatchPSize);
+}
+
+static inline uint32_t rom_patch_region_r_base(
+    const rom_patch_match_regs_t *match) {
+  HARDENED_CHECK_NE(match, NULL);
+  return match->r_base;
+}
+
+static inline uint32_t rom_patch_region_m_base(
+    const rom_patch_match_regs_t *match) {
+  HARDENED_CHECK_NE(match, NULL);
+  return bitfield_field32_read(match->m_base, kRomPatchMatchMBase);
+}
+
+bool rom_patch_valid(const uintptr_t patch_addr) {
+  if (patch_addr == kRomPatchInvalidAddr || patch_addr < kRomPatchBaseAddr ||
+      patch_addr > kRomPatchMaxAddr) {
+    return false;
+  }
+
+  return rom_patch_size_words(otp_read32(patch_addr)) != 0;
```

Comment:
Should this check for `rom_patch_size_words(otp_read32(patch_addr)) >= ((kRomPatchRegionHeaderSizeBytes + kRomPatchRegionMatchTableSizeBytes + kRomPatchRegionSignatureSizeBytes) << 2)`?

Reply:
Yes. This is checked now.